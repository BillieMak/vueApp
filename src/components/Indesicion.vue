<template>
    <img v-if="img" :src="img" alt="Bg">
    <div class="bg-dark">

    </div>

    <div class="indecision-container">
        <input type="text" placeholder="Hasme una Pregunta" v-model="question">
        <p>Recuerda terminar con un signo de interrogacion (?)</p>
        <div v-if="isValidQuestion">
            <h2>{{ answerClass }}</h2>
            <h2>{{ answer }}</h2>
            <!-- <h1>Si, No Pensando</h1> -->
        </div>
    </div>
</template>

<script>
export default {
    name: "Indecision",
    data() {
        return {
            question: "",
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },
    methods: {
        async getAnswer() {
            try {
                this.answer = "Pensando..."
                const { answer, image } = await fetch("https://yesno.wtf/api")
                    .then(res => res.json())
                this.answer = answer
                this.img = image
            } catch (error) {
                this.answer = "No se pudo cargar la respuesta ðŸ¥º"
                this.img = null
            }
        }
    },
    computed: {
        answerClass() {
            return this.answer === "yes" ? "si" : "no"
        }
    },
    watch: {
        question(newQuestion, oldQuestion) {

            this.isValidQuestion = false

            console.log({ newQuestion })

            if (!newQuestion.includes("?")) return
            //TODO: hacer peticion http://yesno.wtf/api
            this.isValidQuestion = true
            this.getAnswer()
        }
    }
}
</script>
<style scoped>
img,
.bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
    position: relative;
    z-index: 99;
}

input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
    margin-bottom: 10px;
}

input:focus {
    outline: none;
}

p {
    color: white;
    font-size: 20px;
    margin-top: 10px;
}

h1,
h2 {
    color: white;
}

h2 {
    margin-top: 150px;
}
</style>