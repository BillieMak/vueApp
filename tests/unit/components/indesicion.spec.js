import { shallowMount } from "@vue/test-utils";
import Indesicion from "@/components/Indesicion";

describe("Indesicion Component", () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indesicion);
    clgSpy = jest.spyOn(console, "log");

    jest.clearAllMocks();
  });

  test("Debe hacer un match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("escribir en el input no debe de disparar nada (console.log)", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    // console.log(wrapper.vm.$data);

    const input = wrapper.find("input");
    await input.setValue("Hola Mundo");

    expect(clgSpy).toHaveBeenCalledTimes(1);

    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test('escribir el simbolo de "?" debe disparar el getAnswer', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    const input = wrapper.find("input");
    await input.setValue("Hola Mundo?");

    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("pruebas en getAnswer", async () => {
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");

    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("yes");
  });

  test("pruebas en getAnswer - fallo en el API", async () => {

    fetch.mockImplementationOnce(() => Promise.reject("API is down"));
    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");    
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar la respuesta ðŸ¥º");

  });
});
