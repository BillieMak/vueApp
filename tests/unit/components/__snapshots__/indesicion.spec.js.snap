// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indesicion Component Debe hacer un match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hasme una Pregunta">
  <p>Recuerda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
